File: Eunleelee/2024-01/1월 6일/택배_배달과_수거하기.py
def solution(cap, n, deliveries, pickups):
    # 끝에 집부터 배달해야 할 것이 없으면 삭제해준다. 배달할 것이 아예 없을 경우도 고려
    while deliveries and deliveries[-1] == 0:
        deliveries.pop()
    
    # 끝에 집부터 수거해야 할 것이 없으면 삭제해준다. 수거할 것이 아예 없을 경우도 고려
    while pickups and pickups[-1] == 0:
        pickups.pop()

    answer = 0
    while deliveries or pickups:
        answer += max(len(deliveries), len(pickups)) * 2

        deliveries_cap = pickups_cap = cap
        while deliveries:
            if deliveries[-1] <= deliveries_cap:
                deliveries_cap -= deliveries.pop()
            else:
                deliveries[-1] -= deliveries_cap
                break
        while pickups:
            if pickups[-1] <= pickups_cap:
                pickups_cap -= pickups.pop()
            else:
                pickups[-1] -= pickups_cap
                break

    return answer

File: yht0827/1월/1월 6일/택배 배달과 수거하기.java
mport java.util .*;

class Solution {
	public long solution(int cap, int n, int[] d, int[] p) {

		long answer = 0;
		int len = d.length;
		//val, idx
		LinkedList<int[]> l1 = new LinkedList<>();
		LinkedList<int[]> l2 = new LinkedList<>();
		for (int i = 0; i < len; i++) {
			if (d[i] > 0)
				l1.add(new int[] {d[i], i + 1});
			if (p[i] > 0)
				l2.add(new int[] {p[i], i + 1});
		}
		int sum = 0;
		int dist = 0;
		while (!l1.isEmpty() || !l2.isEmpty()) {
			sum = 0;
			dist = 0;
			while (!l1.isEmpty() && sum < cap) {
				int[] c = l1.pollLast();
				dist = Math.max(dist, c[1]);

				if (c[0] + sum <= cap) {
					sum += c[0];
				} else {
					c[0] -= (cap - sum);
					l1.add(c);
					break;
				}
			}
			sum = 0;
			while (!l2.isEmpty() && sum < cap) {
				int[] c = l2.pollLast();
				dist = Math.max(dist, c[1]);

				if (c[0] + sum <= cap) {
					sum += c[0];
				} else {
					c[0] -= (cap - sum);
					l2.add(c);
					break;
				}
			}
			answer += dist * 2;
		}

		return answer;
	}
}